File: packages/@cos/shared/src/kafka/metrics.ts
// Prometheus metrics for Kafka producer, consumer, and DLQ — Phase 8
// Metric names aligned to QM-8 required metrics:
//   kafka_messages_produced_total
//   kafka_messages_consumed_total
//   kafka_consumer_lag (gauge)
//   kafka_dlq_depth (gauge — alert when > 0)
//   kafka_producer_error_total

import { Counter, Gauge, Registry } from 'prom-client';

let registry: Registry | null = null;
let messagesProduced: Counter | null = null;
let messagesConsumed: Counter | null = null;
let producerErrors: Counter | null = null;
let consumerLag: Gauge | null = null;
let dlqDepth: Gauge | null = null;

export function initKafkaMetrics(reg: Registry): void {
  registry = reg;

  messagesProduced = new Counter({
    name: 'kafka_messages_produced_total',
    help: 'Total Kafka messages successfully produced',
    labelNames: ['topic', 'event_type'],
    registers: [reg],
  });

  messagesConsumed = new Counter({
    name: 'kafka_messages_consumed_total',
    help: 'Total Kafka messages successfully consumed',
    labelNames: ['topic', 'consumer_group', 'event_type'],
    registers: [reg],
  });

  producerErrors = new Counter({
    name: 'kafka_producer_error_total',
    help: 'Total Kafka producer errors',
    labelNames: ['topic'],
    registers: [reg],
  });

  consumerLag = new Gauge({
    name: 'kafka_consumer_lag',
    help: 'Consumer group lag (number of messages behind)',
    labelNames: ['topic', 'consumer_group', 'partition'],
    registers: [reg],
  });

  dlqDepth = new Gauge({
    name: 'kafka_dlq_depth',
    help: 'Number of messages in DLQ topics (alert when > 0)',
    labelNames: ['dlq_topic'],
    registers: [reg],
  });
}

export function recordProduced(topic: string, eventType: string): void {
  messagesProduced?.labels(topic, eventType).inc();
}

export function recordConsumed(topic: string, groupId: string, eventType: string): void {
  messagesConsumed?.labels(topic, groupId, eventType).inc();
}

export function recordProducerError(topic: string): void {
  producerErrors?.labels(topic).inc();
}

export function setConsumerLag(topic: string, groupId: string, partition: number, lag: number): void {
  consumerLag?.labels(topic, groupId, String(partition)).set(lag);
}

export function setDlqDepth(dlqTopic: string, depth: number): void {
  dlqDepth?.labels(dlqTopic).set(depth);
}

File: libs/go/coskafka/metrics.go
package coskafka

import "github.com/prometheus/client_golang/prometheus"

// Kafka pipeline counters shared by every Go worker's consume/DLQ path.
//
// They live here, next to the code that increments them (consumer.go, dlq.go), rather than in each
// worker's internal/metrics package: coskafka is a shared module and Go forbids it from importing a
// service's internal/ package, so the counters the shared pipeline touches must be shared too. Each
// worker's internal/metrics re-exports and registers these into its own registry and serves them on
// :9464 — coskafka defines and increments; it never registers or exposes.
//
// Metric names, help text and label sets match packages/@cos/shared/src/kafka/metrics.ts exactly,
// so one Grafana panel and the existing alert rules cover the TypeScript and Go consumers alike.
//
// DELIBERATELY NOT DEFINED HERE:
//   - kafka_consumer_lag  — requires querying group offsets via the admin API, not something the
//     consume loop knows. The TypeScript side publishes it.
//   - kafka_dlq_depth     — a gauge of how many messages are SITTING in a DLQ topic. A producer
//     only knows how many it wrote, which is a different number; reporting one as the other would
//     make the KafkaDLQNonEmpty alert lie. DLQ writes show up as kafka_messages_produced_total on
//     the .dlq topic instead.
var (
	// MessagesConsumed counts records that completed the pipeline successfully.
	MessagesConsumed = prometheus.NewCounterVec(prometheus.CounterOpts{
		Name: "kafka_messages_consumed_total",
		Help: "Total Kafka messages successfully consumed",
	}, []string{"topic", "consumer_group", "event_type"})

	// MessagesProduced counts records a worker wrote — today that is DLQ writes.
	MessagesProduced = prometheus.NewCounterVec(prometheus.CounterOpts{
		Name: "kafka_messages_produced_total",
		Help: "Total Kafka messages successfully produced",
	}, []string{"topic", "event_type"})
)
